File: 0785-is-graph-bipartite/0785-is-graph-bipartite.cpp
class Solution {
private:
    bool checkBipartite(int node, vector<vector<int>> &graph, int color[]){
        queue<int> q;
        q.push(node); color[node] = 0;

        while(!q.empty()){
            int n = q.front(); q.pop();

            for(auto adj : graph[n]){
                if(color[adj] == -1){
                    q.push(adj);
                    color[adj] = !color[n]; 
                } else if(color[adj] == color[n]){
                    return false;
                }
            }
        }

        return true;
    }
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int V = graph.size();

        int color[V];
        fill_n(color, V, -1);

        for(int i = 0; i < V; i++){
            if(color[i] == -1){
                if(!checkBipartite(i, graph, color)) return false;
            }
        }

        return true;
    }
};

File: 0785-is-graph-bipartite/0785-is-graph-bipartite.java
class Solution {
    private boolean checkBipartiteDFS(int node, int color, int[] colors, int[][] adj){
        colors[node] = color;

        for(int adjNode : adj[node]){
            if(colors[adjNode] == -1){
                if(!checkBipartiteDFS(adjNode, 1 - color, colors, adj)) return false;
            } else if(colors[adjNode] == colors[node]) return false;
        }

        return true;
    }

    public boolean isBipartite(int[][] adj) {
        int n = adj.length;

        int[] visited = new int[n];
        Arrays.fill(visited, -1);
        

        for(int i = 0; i < n; i++){
            if(visited[i] == -1)
                if(!checkBipartiteDFS(i, 0, visited, adj)) return false;
        }

        return true;
    }
}